drivers/adc/ad7091r5/ad7091r5.c

```diff
@@ -0,0 +1,596 @@
+}
+
+/**
+ * Set device channel.
+ * @param dev - The device structure.
+ * @param channel - Channel.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t ad7091r5_set_channel(struct ad7091r5_dev *dev, uint8_t channel)
+{
+	uint16_t foo;
+	int32_t ret;
+
+	if (!dev)
+		return -EINVAL;
+
+	if (channel >= AD7091R5_CHANNEL_NO)
+		return -EINVAL;
+
+	/* AD7091R5_REG_CHANNEL is a 8-bit register */
+	ret = ad7091r5_i2c_reg_write(dev, AD7091R5_REG_CHANNEL, (BIT(channel) << 8));
+	if (ret)
+		return ret;
+
+	/* There is a latency of one conversion before the channel conversion
+	 * sequence is updated */
+	return ad7091r5_i2c_reg_read(dev, AD7091R5_REG_RESULT, &foo);
+}
+
+/**
+ * @brief Initialize GPIO driver handlers for the GPIOs in the system.
+ *        ad7091r5_init() helper function.
+ * @param dev - ad7091r5_dev device handler.
+ * @param is_software - true: Software reset
+ * 		      - false: hardware reset
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t ad7091r5_reset(struct ad7091r5_dev *dev, bool is_software)
+{
+	int32_t ret;
+
+	if (!dev)
+		return -EINVAL;
+
+	if (is_software) {
+		/* Bit is cleared automatically after reset */
+		return ad7091r5_i2c_write_mask(dev, AD7091R5_REG_CONF,
+					       REG_CONF_RESET_MASK,
+					       REG_CONF_RESET(1));
+	} else {
+		ret = gpio_set_value(dev->gpio_resetn, GPIO_LOW);
+		if (ret < 0)
+			return ret;
+
+		/* reset pulse width, at least 10 ns*/
+		udelay(1);
+		return gpio_set_value(dev->gpio_resetn, GPIO_HIGH);
+	}
+}
+
+/**
+ * @brief Initialize GPIO driver handlers for the GPIOs in the system.
+ *        ad7091r5_init() helper function.
+ * @param dev - ad7091r5_dev device handler.
+ * @param init_param - Pointer to the initialization structure.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+static int32_t ad7091r5_init_gpio(struct ad7091r5_dev *dev,
+				  struct ad7091r5_init_param *init_param)
+{
+	int32_t ret;
+
+	if (!dev || !init_param)
+		return -EINVAL;
+
+	ret = gpio_get_optional(&dev->gpio_resetn, init_param->gpio_resetn);
+	if (ret < 0)
+		return ret;
+
+	ret = gpio_direction_output(dev->gpio_resetn, GPIO_LOW);
+	if (ret < 0)
+		return ret;
+
+	/** Reset to configure pins */
+	return ad7091r5_reset(dev, init_param->gpio_resetn);
+}
+
+/**
+ * Read one sample.
+ * @param dev - The device structure.
+ * @param channel - Channel.
+ * @param read_val - Value.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t ad7091r5_read_one(struct ad7091r5_dev *dev,
+			  uint8_t channel, uint16_t *read_val)
+{
+	uint16_t val;
+	int32_t ret;
+
+	if (!dev || !read_val)
+		return -EINVAL;
+
+	if (channel >= AD7091R5_CHANNEL_NO)
+		return -EINVAL;
+
+	ret = ad7091r5_set_channel(dev, channel);
+	if (ret)
+		return ret;
+
+	ret = ad7091r5_i2c_reg_read(dev, AD7091R5_REG_RESULT, &val);
+	if (ret)
+		return ret;
+
+	if (REG_RESULT_CH_ID(val) != channel)
+		return FAILURE;
```

Comment:
this FAILURE seems to be the only one remaining, could be -EIO or something like that... 
I'm approving since I have nothing else to add